socorro/external/rabbitmq/crashqueue.py

```diff
@@ -0,0 +1,167 @@
+        from_string_converter=class_converter,
+        reference_value_from='resource.rabbitmq',
+    )
+
+    def __init__(self, config, namespace=None, quit_check_callback=None):
+        self.config = config
+        self.namespace = namespace
+        self.quit_check_callback = quit_check_callback
+
+        self.logger = logging.getLogger(__name__ + '.' + self.__class__.__name__)
+
+        # Note: this may continue to grow if we aren't acking certain UUIDs.
+        # We should find a way to time out UUIDs after a certain time.
+        self.acknowledgement_token_cache = {}
+        self.acknowledgment_queue = Queue()
+
+        self.rabbitmq = config.rabbitmq_class(config)
+        self._basic_properties = pika.BasicProperties(
+            # Make message persistent
+            delivery_mode=2,
+        )
+
+    def close(self):
+        pass
+
+    def _basic_get(self, conn, queue):
+        return conn.channel.basic_get(queue=queue)
+
+    def __iter__(self):
+        """Return an iterator over crashes from RabbitMQ.
+
+        Each crash is a tuple of the ``(args, kwargs)`` variety. The lone arg
+        is a crash ID, and the kwargs contain only a callback function which
+        the FTS app will call to send an ack to Rabbit after processing is
+        complete.
+
+        """
+        self._consume_acknowledgement_queue()
+
+        queues = [
+            self.rabbitmq.config.priority_queue_name,
+            self.rabbitmq.config.standard_queue_name,
+            self.rabbitmq.config.reprocessing_queue_name,
+            self.rabbitmq.config.priority_queue_name,
+        ]
+        while True:
+            for queue in queues:
+                method_frame, header_frame, body = retry(
+                    connection_context=self.rabbitmq,
+                    quit_check=self.quit_check_callback,
+                    fun=self._basic_get,
+                    queue=queue
+                )
+                # The body is always a string, so convert it to a string
+                if body:
+                    body = body.decode('utf-8')
+
+                if method_frame and self._suppress_duplicate_jobs(body, method_frame):
+                    continue
+                if method_frame:
+                    break
+            # must consume ack queue before testing for end of iterator
+            # or the last job won't get ack'd
+            self._consume_acknowledgement_queue()
+            if not method_frame:
+                # there was nothing in the queue - leave the iterator
+                return
+            self.acknowledgement_token_cache[body] = method_frame
+            yield (
+                (body,),
+                {'finished_func': partial(self.ack_crash, body)}
+            )
+            queues.reverse()
+
+    def ack_crash(self, crash_id):
+        self.acknowledgment_queue.put(crash_id)
+
+    def _suppress_duplicate_jobs(self, crash_id, acknowledgement_token):
```

Comment:
I think we should nix this code. We have multiple processor nodes, so the idea of "duplicate jobs" really needs to be global across the cluster and not local to a single node. I checked the logs on stage and prod and see one instance of "duplicate job:" in the last 6 days.

Duplicate jobs occur when someone triggers a priority processing or we're doing multiple rounds of reprocessing. The latter would be entirely my fault. Given the size of the priority queue most of the time, I don't think it happens much.

I'll leave it for now since this code isn't going to last long anyhow, but I don't think I'll have this in the Pub/Sub implementation.